lib/dao.js

```diff
@@ -2421,6 +2421,108 @@ DataAccessObject.updateAll = function(where, data, options, cb) {
   return cb.promise;
 };
 
+DataAccessObject.patchById = function patchById(id, data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+  assert(arguments.length >= 3, 'At least three arguments are required');
+
+  var Model = this;
+  var where = {};
+  var idPropertyName = idName(Model);
+
+  where[idPropertyName] = id;
+  if (cb === undefined) {
+    if (typeof options === 'function') {
+      // patchById(id, data, cb)
+      cb = options;
+      options = {};
+    }
+  }
+
+  data = data || {};
+  cb = cb || utils.createPromiseCallback();
+  options = options || {};
+
+  assert((typeof data === 'object') && (data !== null),
+    'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var connector = Model.getDataSource().connector;
+  assert(typeof connector.update === 'function',
+    'update() must be implemented by the connector');
+
+  var hookState = {};
+  this.applyProperties(data);
+
+  var context = {
+    Model: Model,
+    instanceId: id,
+    data: data,
+    hookState: hookState,
+    options: options,
+  };
+
+  Model.notifyObserversOf('before save', context, function(err, ctx) {
+    if (err) return cb(err);
+
+    doUpdate(ctx.instanceId, ctx.data);
+  });
+  function doUpdate(instanceId, data) {
+    try {
+      data = removeUndefined(data);
+      data = Model._coerce(data);
+    } catch (err) {
+      return process.nextTick(function() {
+        cb(err);
+      });
+    }
+
+    function patchCallback(err, info) {
+      if (err) return cb(err);
+      var ctx = {
+        Model: Model,
+        data: data,
+        hookState: hookState,
+        options: options,
+      };
+      Model.notifyObserversOf('loaded', ctx, function(err) {
+        if (err) return cb(err);
+        var context = {
+          Model: Model,
+          isNewInstance: false,
+          hookState: hookState,
+          options: options,
+        };
+        Model.notifyObserversOf('after save', context, function(err) {
+          if (!err) Model.emit('changed', info);
+          cb(err, info);
+        });
+      });
+    }
+
+    var ctx = {
+      Model: Model,
+      instanceId: instanceId,
+      data: data,
+      isNewInstance: false,
+      hookState: hookState,
+      options: options,
+    };
+    Model.notifyObserversOf('persist', ctx, function(err) {
+      if (err) return cb (err);
+      if (connector.update.length === 5) {
+        connector.update(Model.modelName, where, data, options, patchCallback);
```

Comment:
I think we should call `connector.updateAttributes` here, so that we get back full instance data (see the comment above).

`connector.update` could be used in the scenario when a hook changes `ctx.where`, but let's leave that out of scope of this PR.